Plans Next: Fix comparison grid stretching infinitely

chriskmnds opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Proposed Changes

Since merging https://github.com/Automattic/wp-calypso/pull/87003, the `ComparisonGrid` behaves more "fluidly" than expected i.e. it will stretch infinitely unless controlled via consuming-end media queries & max-width.

Controlling this behavior via media queries will probably needs knowledge of the grid-size (and/or other variables, like visible plans) from outside. It's one way, although feels a bit like extra complexity/layers towards reusing this component elsewhere.

The approach here, encodes max-width on the various elements in the grid and on the larger Grid container, which becomes a calculation of "plan columns + header column". The concept of "max column width" is something we have already for the `FeaturesGrid`, so this feels aligned.

We can adjust these constants easily and also move them to a better place later. I think they are only really meaningful in smaller/fewer-plan grids (2 or 3).

## Media

In `/start/hosting` (or `/setup/new-hosted-site`) flow:

**Before**

<img width="700" alt="Screenshot 2024-02-07 at 5 11 02 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/dba25f4d-2ba0-4fd1-8d38-6ed5a1133bdc">


**After**

<img width="700" alt="Screenshot 2024-02-07 at 5 12 21 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/a682cc51-8c71-41ae-a750-b151f4d0d2d7">


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Confirm the comparison-grid in `/start/plans` and `/plans/[ any site ]` renders the correct number of plans on various screen sizes (it should render 2, 3, 4, 4+ on resizing)
- Confirm the grid doesn't keep expanding to eternity
- Confirm a comparison grid with fewer plans does not stretch to eternity. Confirm across all of `/plans/[ newsletter site ]`, `/plans/[ wooexpress site ]` and `/start/hosting` (testing the hosting flow brought this issue to the surface as it renders only 2 plans)
- Confirm borders, highlights, & sticky headers render fine

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

chriskmnds committed:
Plans Next: Fix comparison grid stretching infinitely

jeyip commented:
## Testing
### Requirements

For each of the flows listed below:
- Confirm the comparison-grid in /start/plans and /plans/[ any site ] renders the correct number of plans on various screen sizes (it should render 2, 3, 4, 4+ on resizing)
- Confirm the grid doesn't keep expanding to eternity
- Confirm a comparison grid with fewer plans does not stretch to eternity. Confirm across all of /plans/[ newsletter site ], /plans/[ wooexpress site ] and /start/hosting (testing the hosting flow brought this issue to the surface as it renders only 2 plans)
- Confirm borders, highlights, & sticky headers render fine

Flows:
- [ ] /start
- [ ] /start/hosting
- [ ] /setup/newsletter
- [ ] /setup/wooexpress
- [ ] /plans

### Browsers
- [x] Chrome

oswian reviewed (approved):
Tests well for me  👍

chriskmnds committed:
convert wrapper to flex. remove max-width on container. keep max-width for cells

chriskmnds commented:
Thank you both for testing! In last commit, I removed the max-width on the Grid container and set the display to `flex` on the wrapper. This also works, and probably just `flex` on wrapper was enough (but keeping `max-width` on cells to be safe, with a comment in the code).

Also removed some other properties like `justify-content: space-between` and `align-items: stretch` on the `Row` that don't seem to have any effect now.

I will confirm with production and ship once/if confident.

chriskmnds commented:
> @chriskmnds I noticed that the plans don't render in the 2, 3, 4, 4+ format that you describe when resizing the window ( unsure if this is a big deal though, since this behavior is also what exists in production )

Hey @jeyip - only related to the comparison grid. no changes to the top table :-)

chriskmnds committed:
Revert "convert wrapper to flex. remove max-width on container. keep max-width for cells"

This reverts commit e292b7fb0cff1d124a531e78159a00f4b4067be5.

chriskmnds commented:
> Thank you both for testing! In last commit, I removed the max-width on the Grid container and set the display to `flex` on the wrapper. This also works, and probably just `flex` on wrapper was enough (but keeping `max-width` on cells to be safe, with a comment in the code).
> 
> Also removed some other properties like `justify-content: space-between` and `align-items: stretch` on the `Row` that don't seem to have any effect now.
> 
> I will confirm with production and ship once/if confident.

ehm no. noticing other weirdness - footer taking up room on the side when in narrower view. reverting this last change.

chriskmnds commented:
> /start/hosting ( there is no comparison grid for the plans page in this flow )


@jeyip  it should direct you to `setup/new-hosted-site` once you click "create a site". That's where the 2-plans mix appears with a comparison grid at the bottom. This is the flow that surfaced the issue to me - it was too obvious here 🤷‍♂️

jeyip commented:
> @jeyip it should direct you to setup/new-hosted-site once you click "create a site". That's where the 2-plans mix appears with a comparison grid at the bottom. This is the flow that surfaced the issue to me - it was too obvious here 🤷‍♂️

Ah that's it. Thanks for clarifying!
